### DISCUSSION TITLE
Allow shorthand syntax for single-element tuples

### DISCUSSION THREAD

**Discussion by jez9999** [dotnet/csharplang#6636] [Language Ideas]
I like to use this "single ValueTuple input param" syntax when defining callback `Func` arguments so I can name the input parameters:

```
public void DoThing(int id, Func<(int Id1, int Id2, string Message), bool> fnProcessCallback) {
    //...
    fnProcessCallback(1, 2, "hw");
}

//...

foo.DoThing(12345, (args) => {
    var (id1, id2, msg) = args;
}
```

This works fine until I only want one input param.  Unfortunately I can't just use this syntax to have `ValueTuple`s with names that only have one entry:

```
Func<(int Id1), bool>
```

Could C# support shorthand value tuples with a single element to support this use-case (and maybe a few others)?  Idea for disambiguation - allow a trailing comma in tuple definitions.  A two-, single- and zero-element tuple could then be:

```
(int Id1, int Id2)
(int Id1,)
(,)
```

**Comment by HaloFour** [comment]
See: #883

IMO, this use case is pretty weak.  The names of the tuple don't affect documentation nor do they affect the names of the variables used for deconstruction.  You're better off just using the version of `Func<...>` with the proper arity, or declaring your own custom delegate type.  The latter gives you the ability to provide proper documentation.

**Reply by jnm2** [reply]
> The names of the tuple don't affect documentation

Can you explain more? Tuple element names show up in intellisense and docs.microsoft.com, for instance.

**Reply by HaloFour** [reply]
You can't reference them in xmldocs, though, right?

**Reply by jnm2** [reply]
Not that I know of. If you do want documentation that references delegate arguments at all, you're not using `Func<...>`, right? You'd be declaring a custom delegate?

**Reply by jnm2** [reply]
(And then that delegate documentation doesn't show up when using a method parameter typed as `Func<...>`, does it?)

**Reply by HaloFour** [reply]
That's what I was thinking of re documentation.  I do see that Intellisense could reference other metadata, e.g. `TupleElementNamesAttribute`.

**Comment by jnm2** [comment]
Also see https://github.com/dotnet/csharplang/discussions/5706 and its linked discussions for this particular use case.
